### PR TITLE
IGNITE-21551 Sql. Enable timestamp with local timezone

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-21551

* Added ability to set client time zone in session.
* Adjusted time functions (e.g. current_time, etc) to use the client's time zone.
* Enabled TIMESTAMP_WITH_LOCAL_TIME_ZONE (TIME_WITH_LOCAL_TIME_ZONE still disabled for using in DML [IGNITE-21551](https://issues.apache.org/jira/browse/IGNITE-21551))
* `Literal` `TIMESTAMP_WITH_LOCAL_TIME_ZONE` value adjusted to client time zone.
* Interval tests refactored a bit (basic cases like `select TEMPORAL_TYPE literal +/- interval N unit` moved to single test `ItInterval#testBasicDateTimeIntervalArithmetic`

More detailed explanation

`TIMESTAMP_WITH_LOCAL_TIME_ZONE` type is internally mapped to `Instant`. `Instant` is a UTC timestamp value (`Instan.now()` for example), so the user (if he needs), will display this UTC time point as a value in the local time zone  (for example, `ZonedDateTime.ofInstant(Instant.now(), ZoneId.systemDefault())`). Since we will be reading the same point of time that may be correctly represented in any time zone, we don't need to change anything when reading this value (in client which does not change the return type). Inserting/updating using dynamic parameter also works correctly.

But when the user inserts "literal" value '2021-01-01 10:00:00', he obviously expects this value to be in the local time zone. So, the resulting "literal" value (long) is adjusted taking into account the session time zone (see changes in `RexToLuxTranslator#translateLiteral`).

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/ExpressionFactoryImpl.java [hunk 1]
```diff
@@ -316,6 +318,11 @@ public Iterable<RowT> values(List<RexLiteral> values, RelDataType rowType) {
                 RexLiteral literal = values.get(i);
                 Object val = literal.getValueAs(types.get(field));
 
+                // Literal was parsed as UTC timestamp, now we need to adjust it to the client's time zone.
+                if (val != null && literal.getTypeName() == SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE) {
+                    val = IgniteSqlFunctions.localTimestampToUtc((Long) val, (TimeZone) ctx.get(Variable.TIME_ZONE.camelName));
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/ExpressionFactoryImpl.java]: The value was already boxed)

Changed this casting to `long` for clarity.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/IgniteMethod.java [hunk 2]
```diff
@@ -72,6 +73,9 @@ public enum IgniteMethod {
 
     STRING_TO_TIMESTAMP(IgniteSqlFunctions.class, "timestampStringToNumeric", String.class),
 
+    /** See {@link IgniteSqlFunctions#localTimestampToUtc(Long, TimeZone)}. **/
+    LOCAL_TIMESTAMP_TO_UTC(IgniteSqlFunctions.class, "localTimestampToUtc", Long.class, TimeZone.class),
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/IgniteMethod.java]: Fixed, thanks.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java [hunk 3]
```diff
@@ -564,6 +556,14 @@ public static Object consumeFirstArgument(Object args0, Object args1) {
         return args1;
     }
 
+    /** Returns the timestamp value minus the offset of the specified timezone. */
+    public static Long localTimestampToUtc(Long timestamp, TimeZone timeZone) {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java]: The idea behind this naming (`localTimestampToUtc`) was that here we have a point in time represented in the user's local time zone and we converting it back to `utc` using provided timezone. 

Anyway, so far renamed to `subtractTimeZoneOffset` :thinking:

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java [hunk 4]
```diff
@@ -285,8 +285,7 @@ private static boolean hasConvertableFields(RelDataType resultType) {
         } else if (storageType == LocalTime.class && val instanceof Integer) {
             return LocalTime.ofNanoOfDay(TimeUnit.MILLISECONDS.toNanos(Long.valueOf((Integer) val)));
         } else if (storageType == LocalDateTime.class && (val instanceof Long)) {
-            return LocalDateTime.ofEpochSecond(TimeUnit.MILLISECONDS.toSeconds((Long) val),
-                    (int) TimeUnit.MILLISECONDS.toNanos((Long) val % 1000), ZoneOffset.UTC);
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java:289]: This computation is incorrect for negative `val`.
Example:
```
SELECT timestamp '1969-01-01 02:30:00.123'
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java [hunk 5]
```diff
@@ -285,8 +285,7 @@ private static boolean hasConvertableFields(RelDataType resultType) {
...
+            return LocalDateTime.ofInstant(Instant.ofEpochMilli((long) val), ZoneOffset.UTC);
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java:288]: Computation of nanoseconds was incorrect for negative `val`.
```
SELECT timestamp '1969-01-01 02:30:00.123'
```

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java [hunk 6]
```diff
@@ -197,45 +200,127 @@ public void testDateTime() {
...
                 .returns(sqlDateTime("1992-09-27T03:30:00.000")).check();
+        assertQuery("SELECT timestamp with local time zone '1992-09-26 02:30:00'")
+                .returns(sqlTimestamp("1992-09-26T02:30:00.000")).check();
         assertQuery("SELECT timestamp '1992-09-26 02:30:00' + interval (23) hours")
                 .returns(sqlDateTime("1992-09-27T01:30:00.000")).check();
+        assertQuery("SELECT timestamp with local time zone '1992-09-26 02:30:00' + interval (23) hours")
+                .returns(sqlTimestamp("1992-09-27T01:30:00.000")).check();
         assertQuery("SELECT timestamp '1992-09-26 02:30:00' + interval (24) hours")
                 .returns(sqlDateTime("1992-09-27T02:30:00.000")).check();
+        assertQuery("SELECT timestamp with local time zone '1992-09-26 02:30:00' + interval (24) hours")
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java]: Interval tests refactored a bit (basic cases like `select TEMPORAL_TYPE literal +/- interval N unit` moved to single test `ItInterval#testBasicDateTimeIntervalArithmetic`

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlDdlParserTest.java [hunk 7]
```diff
@@ -631,18 +630,14 @@ public void dropIndexIfExists() {
     /**
      * Ensures that the user cannot use the TIME_WITH_LOCAL_TIME_ZONE and TIMESTAMP_WITH_LOCAL_TIME_ZONE types for table columns.
      */
-    // TODO: Remove after https://issues.apache.org/jira/browse/IGNITE-19274 is implemented.
+    // TODO: Remove after https://issues.apache.org/jira/browse/IGNITE-21555 is implemented.
     @Test
     public void timestampWithLocalTimeZoneIsNotSupported() {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlDdlParserTest.java]: Fixed, thanks

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionContext.java [hunk 8]
```diff
@@ -115,14 +118,15 @@ public ExecutionContext(
         this.localNode = localNode;
         this.originatingNodeName = originatingNodeName;
         this.txAttributes = txAttributes;
+        this.timeZoneId = Objects.requireNonNullElseGet(timeZoneId, ZoneId::systemDefault);
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionContext.java]: Done, thanks

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryProperty.java [hunk 9]
```diff
@@ -32,6 +33,7 @@ public final class QueryProperty {
     public static final Property<Set<SqlQueryType>> ALLOWED_QUERY_TYPES =
             new Property<>("allowed_query_types", cast(Set.class));
     public static final Property<String> DEFAULT_SCHEMA = new Property<>("default_schema", String.class);
+    public static final Property<ZoneId> DEFAULT_TIME_ZONE_ID = new Property<>("default_time_zone_id", ZoneId.class);
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryProperty.java]: Fixed, thanks

**Review by korlov42 — APPROVED** [review]

**Pull request merged**